rfc/0001-crashtracker-user-visable-log.md

```diff
@@ -0,0 +1,703 @@
+Having a single structured format allows us to work across languages.
+
+## Proposed format
+A natural language description of the proposed json format is given here.
+An example is given in Appendix A, and the schema is given in Appendix B.
+
+### Required fields
+- `errortype`:
+    Currently, the only value is "crash", but this allows for extension to capture unhandled soft-errors, e.g. "panic", "uncaught exception", etc.
+- `incomplete`:
+    Boolean `false` if the crashreport is complete (i.e. contains all intended data), `true` if there is expected missing data.
+    This can happen becasue the crashtracker is architected to stream data to an out of process receiver, allowing a partial crash report to be emitted even in the case where the crashtracker itself crashed during stack trace collection.
+    This MUST be set to `true` if any required field is missing.
+- `metadata`:
+    Metadata about the system in which the crash occurred:
+    - `library_name`
+    - `library_version`
+    - `family`
+    - `tags`:
+      A set of key:value pairs, representing any tags the crashtracking system wishes to associate with the crash.
+      Examples would include "hostname", "service", and any configuration information the system wishes to track.
+- `os_info`: 
+    The OS + processor architecture on which the crash occurred.
+- `stacktrace`: 
+    This represents the stack of the crashing thread.
+    See below for more details on how stacktraces are formatted.
+- `timestamp`:
+    The time at which the crash occurred, in ISO 8601 format.
+- `uuid`:
+    A UUID which uniquely identifies the crash.
+- `version_id`:
+    A Semver compatible ID for this format. [TODO, should it be semver?]
+
+### Optional fields
+Any field not listed as "Required" is optional.
+In order to minimize logging overhead, producers SHOULD NOT emit anything for an optional field.
+Consumers MUST accept json with elided optional fields.
+
+- `additional_stacktraces`:
+    This field contains a `Map<ThreadId, Stacktrace>`.
+    In a multi-threaded program, the collector SHOULD collect the stacktraces of all active threads, and report them here.
+- `counters`:
+    The crashtracker offers a mechanism for programs to register counters to track which operations were active at the time of the crash.
+    At present, this is only used by the profiler, but this may be extended in the future.
+- `files`:
+    The collector MAY collect useful files, such as `/proc/self/maps` or `/proc/meminfo`, and include them here.
+    Files are stored as an array of plain text strings, one per line.
+- `proc_info`: 
+    Currently, just tracks the PID of the crashing process.  
+    In the future, this may record additional info about the crashing process.
+- `siginfo`:
+    The name and signal number of the crashing signal (on UNIX systems)
```

Comment:
I wonder if we should extend this. Signal is a very unix-specific concept. Applications can also crash because of an unhandled exception, which we have no normalized way to store right now (in .NET I'm currently storing that as a tag)

Reply:
Extended.